File: frontend/src/services/rebalance.ts
export type RebalanceBroker = 'zerodha' | 'angelone' | 'both'
export type RebalanceMode = 'MANUAL' | 'AUTO'
export type RebalanceTargetKind = 'GROUP' | 'HOLDINGS'
export type RebalanceMethod = 'TARGET_WEIGHT' | 'SIGNAL_ROTATION' | 'RISK_PARITY'
export type RebalanceRotationWeighting = 'EQUAL' | 'SCORE' | 'RANK'

export type RebalanceRotationConfig = {
  signal_strategy_version_id: number
  signal_output: string
  signal_params?: Record<string, unknown>
  universe_group_id?: number | null
  screener_run_id?: number | null
  top_n: number
  weighting: RebalanceRotationWeighting
  sell_not_in_top_n: boolean
  min_price?: number | null
  min_avg_volume_20d?: number | null
  symbol_whitelist?: string[]
  symbol_blacklist?: string[]
  require_positive_score: boolean
}

export type RebalanceRiskWindow = '6M' | '1Y'

export type RebalanceRiskConfig = {
  window: RebalanceRiskWindow
  timeframe: '1d'
  min_observations: number
  min_weight: number
  max_weight: number
  max_iter: number
  tol: number
}

export type RebalanceTrade = {
  symbol: string
  exchange?: string | null
  side: 'BUY' | 'SELL'
  qty: number
  estimated_price: number
  estimated_notional: number
  target_weight: number
  live_weight: number
  drift: number
  current_value: number
  desired_value: number
  delta_value: number
  scale: number
  reason?: Record<string, unknown>
}

export type RebalancePreviewSummary = {
  portfolio_value: number
  budget: number
  scale: number
  total_buy_value: number
  total_sell_value: number
  turnover_pct: number
  budget_used: number
  budget_used_pct: number
  max_abs_drift_before: number
  max_abs_drift_after: number
  trades_count: number
}

export type RebalancePreviewResult = {
  target_kind: RebalanceTargetKind
  group_id?: number | null
  broker_name: 'zerodha' | 'angelone'
  trades: RebalanceTrade[]
  summary: RebalancePreviewSummary
  warnings: string[]
  derived_targets?: Array<Record<string, unknown>> | null
}

export async function previewRebalance(payload: {
  target_kind: RebalanceTargetKind
  group_id?: number | null
  broker_name: RebalanceBroker
  rebalance_method?: RebalanceMethod | null
  rotation?: RebalanceRotationConfig | null
  risk?: RebalanceRiskConfig | null
  budget_pct?: number | null
  budget_amount?: number | null
  drift_band_abs_pct?: number | null
  drift_band_rel_pct?: number | null
  max_trades?: number | null
  min_trade_value?: number | null
}): Promise<RebalancePreviewResult[]> {
  const res = await fetch('/api/rebalance/preview', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(payload),
  })
  if (!res.ok) {
    const body = await res.text()
    throw new Error(
      `Failed to preview rebalance (${res.status})${body ? `: ${body}` : ''}`,
    )
  }
  const data = (await res.json()) as { results: RebalancePreviewResult[] }
  return data.results ?? []
}

export type RebalanceRunOrder = {
  id: number
  run_id: number
  order_id?: number | null
  symbol: string
  exchange?: string | null
  side: string
  qty: number
  estimated_price?: number | null
  estimated_notional?: number | null
  status: string
  created_at: string
}

export type RebalanceRun = {
  id: number
  owner_id?: number | null
  group_id: number
  broker_name: string
  status: string
  mode: string
  idempotency_key?: string | null
  summary_json?: string | null
  error_message?: string | null
  created_at: string
  executed_at?: string | null
  orders: RebalanceRunOrder[]
}

export async function executeRebalance(payload: {
  target_kind: RebalanceTargetKind
  group_id?: number | null
  broker_name: RebalanceBroker
  rebalance_method?: RebalanceMethod | null
  rotation?: RebalanceRotationConfig | null
  risk?: RebalanceRiskConfig | null
  budget_pct?: number | null
  budget_amount?: number | null
  drift_band_abs_pct?: number | null
  drift_band_rel_pct?: number | null
  max_trades?: number | null
  min_trade_value?: number | null
  mode: RebalanceMode
  execution_target: 'LIVE' | 'PAPER'
  order_type: 'MARKET' | 'LIMIT'
  product: 'CNC' | 'MIS'
  idempotency_key?: string | null
}): Promise<Array<{ run: RebalanceRun | null; created_order_ids: number[] }>> {
  const res = await fetch('/api/rebalance/execute', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(payload),
  })
  if (!res.ok) {
    const body = await res.text()
    throw new Error(
      `Failed to execute rebalance (${res.status})${body ? `: ${body}` : ''}`,
    )
  }
  const data = (await res.json()) as {
    results: Array<{ run?: RebalanceRun | null; created_order_ids: number[] }>
  }
  return (data.results ?? []).map((r) => ({
    run: (r.run ?? null) as RebalanceRun | null,
    created_order_ids: r.created_order_ids ?? [],
  }))
}

export async function listRebalanceRuns(params: {
  group_id?: number
  broker_name?: 'zerodha' | 'angelone' | null
}): Promise<RebalanceRun[]> {
  const search = new URLSearchParams()
  if (params.group_id != null) search.set('group_id', String(params.group_id))
  if (params.broker_name) search.set('broker_name', params.broker_name)
  const url = `/api/rebalance/runs${search.toString() ? `?${search}` : ''}`
  const res = await fetch(url)
  if (!res.ok) {
    const body = await res.text()
    throw new Error(
      `Failed to load rebalance history (${res.status})${body ? `: ${body}` : ''}`,
    )
  }
  return (await res.json()) as RebalanceRun[]
}

export async function getRebalanceRun(runId: number): Promise<RebalanceRun> {
  const res = await fetch(`/api/rebalance/runs/${runId}`)
  if (!res.ok) {
    const body = await res.text()
    throw new Error(
      `Failed to load rebalance run (${res.status})${body ? `: ${body}` : ''}`,
    )
  }
  return (await res.json()) as RebalanceRun
}

export type RebalanceScheduleFrequency =
  | 'WEEKLY'
  | 'MONTHLY'
  | 'QUARTERLY'
  | 'CUSTOM_DAYS'

export type RebalanceScheduleRoll = 'NEXT' | 'PREV' | 'NONE'

export type RebalanceScheduleConfig = {
  frequency: RebalanceScheduleFrequency
  time_local: string
  timezone: string
  weekday?: number | null
  day_of_month?: number | 'LAST' | null
  interval_days?: number | null
  roll_to_trading_day: RebalanceScheduleRoll
}

export type RebalanceSchedule = {
  group_id: number
  enabled: boolean
  config: RebalanceScheduleConfig
  next_run_at?: string | null
  last_run_at?: string | null
  updated_at?: string | null
}

export async function fetchRebalanceSchedule(groupId: number): Promise<RebalanceSchedule> {
  const res = await fetch(`/api/rebalance/schedule?group_id=${groupId}`)
  if (!res.ok) {
    const body = await res.text()
    throw new Error(
      `Failed to load rebalance schedule (${res.status})${body ? `: ${body}` : ''}`,
    )
  }
  return (await res.json()) as RebalanceSchedule
}

export async function updateRebalanceSchedule(
  groupId: number,
  payload: {
    enabled?: boolean
    config?: RebalanceScheduleConfig
  },
): Promise<RebalanceSchedule> {
  const res = await fetch(`/api/rebalance/schedule?group_id=${groupId}`, {
    method: 'PUT',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(payload),
  })
  if (!res.ok) {
    const body = await res.text()
    throw new Error(
      `Failed to update rebalance schedule (${res.status})${body ? `: ${body}` : ''}`,
    )
  }
  return (await res.json()) as RebalanceSchedule
}

File: backend/app/schemas/rebalance.py
from __future__ import annotations

from datetime import datetime
from typing import Dict, List, Literal, Optional

from pydantic import BaseModel, Field

from app.pydantic_compat import PYDANTIC_V2, ConfigDict, model_validator

RebalanceBrokerScope = Literal["zerodha", "angelone", "both"]
RebalanceMode = Literal["MANUAL", "AUTO"]
RebalanceTargetKind = Literal["GROUP", "HOLDINGS"]


class RebalancePreviewRequest(BaseModel):
    target_kind: RebalanceTargetKind = "GROUP"
    group_id: Optional[int] = Field(
        None,
        ge=1,
        description="Required when target_kind=GROUP.",
    )
    broker_name: RebalanceBrokerScope = "zerodha"

    budget_pct: Optional[float] = Field(
        0.10,
        ge=0.0,
        le=1.0,
        description="Fraction of portfolio value to rebalance (0.0 to 1.0).",
    )
    budget_amount: Optional[float] = Field(
        None,
        ge=0.0,
        description="Absolute INR budget; overrides budget_pct when set.",
    )

    drift_band_abs_pct: float = Field(
        0.02,
        ge=0.0,
        le=1.0,
        description="Absolute drift band (fraction).",
    )
    drift_band_rel_pct: float = Field(
        0.15,
        ge=0.0,
        le=1.0,
        description="Relative drift band as fraction of target weight.",
    )

    max_trades: int = Field(10, ge=0, le=200)
    min_trade_value: float = Field(2000.0, ge=0.0)

    @model_validator(mode="before")
    def _validate_group_id(cls, values):  # type: ignore[no-untyped-def]
        # Pydantic v1 compatibility: root_validator signature is (cls, values).
        if not isinstance(values, dict):
            return values
        target_kind = values.get("target_kind") or "GROUP"
        group_id = values.get("group_id")
        if str(target_kind).upper() == "GROUP":
            if group_id is None:
                raise ValueError("group_id is required when target_kind=GROUP.")
        return values


class RebalanceTrade(BaseModel):
    symbol: str
    exchange: Optional[str] = None
    side: Literal["BUY", "SELL"]
    qty: int
    estimated_price: float
    estimated_notional: float

    target_weight: float
    live_weight: float
    drift: float
    current_value: float
    desired_value: float
    delta_value: float
    scale: float

    reason: Dict[str, object] = {}


class RebalancePreviewSummary(BaseModel):
    portfolio_value: float
    budget: float
    scale: float

    total_buy_value: float
    total_sell_value: float
    turnover_pct: float
    budget_used: float
    budget_used_pct: float

    max_abs_drift_before: float
    max_abs_drift_after: float
    trades_count: int


class RebalancePreviewResult(BaseModel):
    target_kind: RebalanceTargetKind = "GROUP"
    group_id: Optional[int] = None
    broker_name: Literal["zerodha", "angelone"]
    trades: List[RebalanceTrade] = []
    summary: RebalancePreviewSummary
    warnings: List[str] = []


class RebalancePreviewResponse(BaseModel):
    results: List[RebalancePreviewResult]


class RebalanceExecuteRequest(RebalancePreviewRequest):
    mode: RebalanceMode = "MANUAL"
    execution_target: Literal["LIVE", "PAPER"] = "LIVE"
    order_type: Literal["MARKET", "LIMIT"] = "MARKET"
    product: Literal["CNC", "MIS"] = "CNC"
    idempotency_key: Optional[str] = Field(None, max_length=128)


class RebalanceRunOrderRead(BaseModel):
    id: int
    run_id: int
    order_id: Optional[int] = None
    symbol: str
    exchange: Optional[str] = None
    side: str
    qty: float
    estimated_price: Optional[float] = None
    estimated_notional: Optional[float] = None
    status: str
    created_at: datetime

    if PYDANTIC_V2:
        model_config = ConfigDict(from_attributes=True)
    else:  # pragma: no cover

        class Config:
            orm_mode = True


class RebalanceRunRead(BaseModel):
    id: int
    owner_id: Optional[int] = None
    group_id: int
    broker_name: str
    status: str
    mode: str
    idempotency_key: Optional[str] = None
    summary_json: Optional[str] = None
    error_message: Optional[str] = None
    created_at: datetime
    executed_at: Optional[datetime] = None
    orders: List[RebalanceRunOrderRead] = []

    if PYDANTIC_V2:
        model_config = ConfigDict(from_attributes=True)
    else:  # pragma: no cover

        class Config:
            orm_mode = True


class RebalanceExecuteResult(BaseModel):
    run: Optional[RebalanceRunRead] = None
    created_order_ids: List[int] = []


class RebalanceExecuteResponse(BaseModel):
    results: List[RebalanceExecuteResult]


__all__ = [
    "RebalanceBrokerScope",
    "RebalanceMode",
    "RebalanceTargetKind",
    "RebalancePreviewRequest",
    "RebalancePreviewResponse",
    "RebalancePreviewResult",
    "RebalanceTrade",
    "RebalancePreviewSummary",
    "RebalanceExecuteRequest",
    "RebalanceExecuteResponse",
    "RebalanceExecuteResult",
    "RebalanceRunRead",
    "RebalanceRunOrderRead",
]
